### PR TITLE
Support for UserInfo response encrypted

### LINKED ISSUE #10517: Support for UserInfo response encrypted
### Description

As of OpenId spec (https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse) UserInfo endpoint response could be produced plain (json format), signed(JWS), encrypted(JWE) and encrypted+signed(JWE containing a JWS); Keycloak only support this spec up to signing (none support for encryption).

### Discussion

https://github.com/keycloak/keycloak/discussions/10194

### Motivation

User Info responses may contains sensitive personal data, so sometimes these responses are required to be encrypted as of openID specification

### Details

_No response_

### PR DESCRIPTION
**Pull request by giacomoa**
Close #10517

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java [hunk 1]
```diff
@@ -25,14 +25,17 @@
 import org.keycloak.common.ClientConnection;
 import org.keycloak.common.VerificationException;
 import org.keycloak.common.constants.ServiceAccountConstants;
-import org.keycloak.crypto.SignatureProvider;
-import org.keycloak.crypto.SignatureSignerContext;
-import org.keycloak.crypto.SignatureVerifierContext;
+import org.keycloak.crypto.*;
```
Comment (tnorimat) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java:28]: By following the keycloak's convention, it might be better not to include wildcard "*" in an import directive.

**Review by tnorimat — CHANGES_REQUESTED** [review]
@giacomoa Thank you for our PR. I've not yet investigated this PR's codes completely but added review comments. Could you check them?

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java (hunk 1 above)
Comment (giacomoa) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java:28]: Hello @tnorimat, thanks for your feedback.
My IDE (IntelliJ IDEA) seems to include wildcard automatically when imported class for a package is >=  5, I see there are other wildcard in the project with same convention (e.g org.keycloak.models.* in org.keycloak.models.cache.infinispan.RealmAdapter, java.util.* in org.keycloak.storage.ldap.idm.query.internal.LDAPQuery).
Let me know if I need to provide a fix anyway or if you have further considerations.

**Comment by giacomoa** [comment]
Hello @tnorimat , do you have any further feedback about this PR?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/OIDCClientRegistrationTest.java [hunk 2]
```diff
@@ -321,6 +321,47 @@ public void testMtlsHoKTokenEnabled() throws Exception {
...
+    public void testUserInfoEncryptedResponse() throws Exception {
+        OIDCClientRepresentation response = null;
+        OIDCClientRepresentation updated = null;
+        try {
+            // create (no specification)
+            OIDCClientRepresentation clientRep = createRep();
+
+            response = reg.oidc().create(clientRep);
+            Assert.assertEquals(Boolean.FALSE, response.getTlsClientCertificateBoundAccessTokens());
+            Assert.assertNotNull(response.getClientSecret());
```
Comment (tnorimat) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/OIDCClientRegistrationTest.java:334]: It might be better to delete this two lines.

**Review by tnorimat — COMMENTED** [review]
@giacomoa I've added some review comments. Could you check them?

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java [hunk 3]
```diff
@@ -245,12 +253,22 @@ private Response issueUserInfo(String tokenString) {
...
 
-            responseBuilder = Response.ok(signedUserInfo).header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JWT);
+            responseBuilder = Response.ok(cfg.isUserInfoEncryptionRequired() ? jweFromContent(signedUserInfo, "JWT") :
+                    signedUserInfo).header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JWT);
 
             event.detail(Details.SIGNATURE_REQUIRED, "true");
             event.detail(Details.SIGNATURE_ALGORITHM, cfg.getUserInfoSignedResponseAlg().toString());
         } else {
-            responseBuilder = Response.ok(claims).header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON);
+            if(cfg.isUserInfoEncryptionRequired()) {
```
[the remark this answers is not in the document]
Comment (giacomoa) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java]: Ok, I've done both in new commit

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java [hunk 4]
```diff
@@ -68,6 +68,26 @@ public boolean isUserInfoSignatureRequired() {
...
+
+    public String getUserInfoEncryptedResponseEnc() {
+        return getAttribute(OIDCConfigAttributes.USER_INFO_ENCRYPTED_RESPONSE_ENC);
+    }
+
+    public void setUserInfoEncryptedResponseEnc(String algorithm) {
+        setAttribute(OIDCConfigAttributes.USER_INFO_ENCRYPTED_RESPONSE_ENC, algorithm);
+    }
+
+    public boolean isUserInfoEncryptionRequired() {
```
[the remark this answers is not in the document]
Comment (giacomoa) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java:87]: You're right, I've modified as you suggested in OIDCAdvancedConfigWrapper and added an overload in DefaultTokenManager to manage a default value for encryption_enc value.
I think also ID Token encryption has the same issue, cause it doesn't default encryption_enc to A128CBC-HS256 in encryption_alg is specified. We should manage this in a different bug, do you agree?

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java [hunk 5]
```diff
@@ -260,6 +278,35 @@ private Response issueUserInfo(String tokenString) {
...
+        JWEAlgorithmProvider jweAlgorithmProvider = cekManagementProvider.jweAlgorithmProvider();
+
+        ContentEncryptionProvider contentEncryptionProvider = session.getProvider(ContentEncryptionProvider.class, encAlgorithm);
+        JWEEncryptionProvider jweEncryptionProvider = contentEncryptionProvider.jweEncryptionProvider();
+
+        ClientModel client = session.getContext().getClient();
+
+        KeyWrapper keyWrapper = PublicKeyStorageManager.getClientPublicKeyWrapper(session, client, JWK.Use.ENCRYPTION, algAlgorithm);
+        if (keyWrapper == null) {
+            throw new RuntimeException("can not get encryption KEK");
```
[the remark this answers is not in the document]
Comment (giacomoa) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java:317]: Ok, done

File: themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties [hunk 6]
```diff
@@ -404,6 +404,10 @@ id-token-encrypted-response-enc=ID Token Encryption Content Encryption Algorithm
...
 user-info-signed-response-alg=User Info Signed Response Algorithm
 user-info-signed-response-alg.tooltip=JWA algorithm used for signed User Info Endpoint response. If set to 'unsigned', User Info Response won't be signed and will be returned in application/json format.
+user-info-encrypted-response-alg=User Info Response Encryption Key Management Algorithm
+user-info-encrypted-response-alg.tooltip=JWA Algorithm used for key management in encrypting User Info Endpoint responses. This option is needed if you want encrypted User Info Endpoint responses. If left empty, User Info Endpoint responses are not encrypted.
+user-info-encrypted-response-enc=User Info Response Encryption Content Encryption Algorithm
+user-info-encrypted-response-enc.tooltip=JWA Algorithm used for content encryption in encrypting User Info Endpoint responses. This option is needed just if you want encrypted User Info Endpoint responses. If left empty, User Info Endpoint responses are not encrypted.
```
[the remark this answers is not in the document]
Comment (giacomoa) [reply themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties]: Updated the tooltip according to specification and new implementation provided

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/OIDCClientRegistrationTest.java (hunk 2 above)
Comment (giacomoa) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/OIDCClientRegistrationTest.java]: Ok, done

**Comment by tnorimat** [comment]
@giacomoa I've missed the following 2 points in my code review. Could you also check them?

1. Server Metadata
The following [OIDC server metadata](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata) should be returned:
- userinfo_encryption_alg_values_supported
- userinfo_encryption_enc_values_supported 
The following PR might be helpful:
- implementation: https://github.com/keycloak/keycloak/pull/6237/files#diff-11c6b1f6947fb113d3e1c41a019acf8ea908c3c7ce9106c42f46af9ce7aee770
- test: https://github.com/keycloak/keycloak/pull/6237/files#diff-50bce047484fc40b152c965bcbb0be2498df27c36a2aed6c28a04eff0a9d24d8

2. Test Patterns
It might be better to test many combinations of `alg` and `enc` as possible.
The following PR might be helpful.
https://github.com/keycloak/keycloak/pull/6237/files#diff-936e7264188091ab29ccb4bfd96d0f3b4b533ea05a6700596f054a75800f73cb

**Comment by tnorimat** [comment]
@giacomoa BTY, do you know use cases that requires UserInfo response encryption? If so, could you tell me them?

**Comment by giacomoa** [comment]
> @giacomoa I've missed the following 2 points in my code review. Could you also check them?
> 
> 1. Server Metadata
>    The following [OIDC server metadata](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata) should be returned:
> 
> * userinfo_encryption_alg_values_supported
> * userinfo_encryption_enc_values_supported
>   The following PR might be helpful:
> * implementation: https://github.com/keycloak/keycloak/pull/6237/files#diff-11c6b1f6947fb113d3e1c41a019acf8ea908c3c7ce9106c42f46af9ce7aee770
> * test: https://github.com/keycloak/keycloak/pull/6237/files#diff-50bce047484fc40b152c965bcbb0be2498df27c36a2aed6c28a04eff0a9d24d8
> 
> 2. Test Patterns
>    It might be better to test many combinations of `alg` and `enc` as possible.
>    The following PR might be helpful.
>    https://github.com/keycloak/keycloak/pull/6237/files#diff-936e7264188091ab29ccb4bfd96d0f3b4b533ea05a6700596f054a75800f73cb
Hi @tnorimat, I've already implemented Server Metadata response in PR provided here: https://github.com/giacomoa/keycloak/blob/feature/user-info-encrypted/services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProvider.java#L145
is there something missing?
For second point I will provide more tests according to PR you linked.

**Comment by giacomoa** [comment]
> @giacomoa BTY, do you know use cases that requires UserInfo response encryption? If so, could you tell me them? I think the motivation of this PR is important.

I added a short motivation in linked issue: https://github.com/keycloak/keycloak/issues/10517
In our company we have requirements to encrypt UserInfo responses for OIDC Client external to our organization to reduce possible sensitive information exposures (e.g in client's server logs)

**Comment by tnorimat** [comment]
@giacomoa I'm sorry, I've missed that you've already supported Server Metadata support in this PR.

**Comment by giacomoa** [comment]
Ok, no problem :) I've updated UserInfoTest to check all combination of alg and enc in last commit.
Let me know if there's something more.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java:87]: Yes. I agree with you.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java [hunk 7]
```diff
@@ -246,6 +262,219 @@ public void testSuccess_postMethod_header_textEntity() throws Exception {
...
 
+
+    @Test
+    public void testSuccessEncryptedAndSignedResponse() throws Exception {
+        testUserInfoSignatureAndEncryption(org.keycloak.crypto.Algorithm.RS256, JWEConstants.RSA_OAEP, JWEConstants.A256GCM);
+    }
+
+    @Test
+    public void testSuccessEncryptedResponseAlgRSA_OAEPEncA256GCM() throws Exception {
+        testUserInfoSignatureAndEncryption(null, JWEConstants.RSA_OAEP, JWEConstants.A256GCM);
```
Comment (tnorimat) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java:273]: 18 of 19 tests set null to "alg" of signing. I think most use cases use signed  and encrypt. It seems good to set some valid signature algorithm to "alg" of signing in 18 of 19 tests and null to "alg" of signing in remaining test.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java [hunk 8]
```diff
@@ -246,6 +262,219 @@ public void testSuccess_postMethod_header_textEntity() throws Exception {
...
+    }
```
Comment (tnorimat) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java:274]: It seems good to add a test that sets null to "enc" to test a default behavior (expects A128CBC-HS256).

**Review by tnorimat — COMMENTED** [review]
@giacomoa I've added some comments.  Could you check them?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java (hunk 8 above)
Comment (giacomoa) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java:274]: Hi @tnorimat , 
about testing various combination of signing and encryption algorithms we have following:

- signing alg: 12 possible algs - 3 HS algs not of interest for our use case (simmetric key not available to clients for signature's check); we'll consider 9 total algs

- encryption alg: 3 possible algs

- encryption enc: 6 possible alg + 1 for default; we'll consider 7 total algs

In tests provided I check all combination of encryption enc and alg (21 tests) and in 18 of them I set also a signing algorithm (each signing alg is used in 2 different test), remaining 3 tests only apply encryption without signing; could be ok?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java (hunk 7 above)
Comment (mposolda) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java]: +1 to this point

Also regarding test coverage, it is great that your PR has good test coverage. On the other hand, 19 new test methods in the UserInfoTest looks like quite a lot to me. Every new test method adds some overhead when running the automated tests... Is it possible to slightly limit this for example to 5-7 methods instead of 19? Ideally including the possibilities for "null" values correctly resolved to expected defaults as Takashi suggested. IMO it is not strictly needed to test every possible combination, but for example just some "representative" subset of them. WDYT?

**Review by mposolda — CHANGES_REQUESTED** [review]
@giacomoa Thanks for the contribution! And thanks @tnorimat for the review!

I have only one small note regarding the count of the tests. But besides that, it looks good to me. I hope we can have this PR quickly resolved, so it can be included in Keycloak 18 release.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java (hunk 8 above)
Comment (tnorimat) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java:274]: @mposolda Yes, I agree with you.
@giacomoa Could you follow @mposolda 's comment?

**Review by mposolda — APPROVED** [review]
@giacomoa Thanks! I am approving your PR.

@tnorimat WDYT? I can merge this once you approve this PR as well (assuming tests will pass).

**Review by tnorimat — APPROVED** [review]

**Review by tnorimat — COMMENTED** [review]
It might be better to squash all 5 commits to 1 commit. AFAIK, 1 commit per PR is keycloak's convention.

**Comment by mposolda** [comment]
@tnorimat True. I did that during the merge.
@giacomoa Thanks again!
